Use arbos state in precompile (#830)
OffchainLabs/arbitrum-docs: arbitrum-docs/launch-orbit-chain/how-tos/customize-precompile.mdx

To learn more about the gas cost model, see [how to estimate gas](/devs-how-tos/how-to-estimate-gas).

## Option 5: Call and modify state

> In this example, we'll demonstrate how to _read from_ and _write to_ a precompile contract's ArbOS state.

Reviewer: We can maybe add a link to `/arbos/#arbos-state` in "ArbOS state" for some extra reference

Author: I didn't apply those tags to there 3 lines is because they point to specific line, so I use a constant url here, if we use it, the file will update to the latest version which might be different line in the future

To learn more about the gas cost model, see [how to estimate gas](/devs-how-tos/how-to-estimate-gas).

> ## Option 5: Call and modify state.

Reviewer 2: ```suggestion
## Option 5: Call and modify state
```

> In this example, we will continue to introduce how to read and write your state in a precompile contract under ArbOS.

Reviewer 2: ```suggestion
In this example, we'll demonstrate how to _read from_ and _write to_ a precompile contract's ArbOS state.
```

> First you need to open the [arbosstate.go][arbosstate_file_link] file, then find the [ArbosState][arbosstate_link] structure in this file. This is where the state stored in arbos is defined.

Reviewer 2: ```suggestion
First, open the [arbosstate.go][arbosstate_file_link] file and locate the [ArbosState][arbosstate_link] structure. This is where ArbOS state is defined.
```

> We can define a state called `myNumber` here, `myNumber` uses the `storage.StorageBackedUint64` type. You can find more related types in [storage.go][storage_link]:

Reviewer 2: ```suggestion
Define a state key called `myNumber` of type `storage.StorageBackedUint64`. You can find more types in [storage.go][storage_link]:
```

_state_ or _state key_, or maybe _state variable_? Is this term formally standardized, or colloquially normalized?

	arbosVersion           uint64                      // version of the ArbOS storage format and semantics
	upgradeVersion         storage.StorageBackedUint64 // version we're planning to upgrade to, or 0 if not planning to upgrade
	upgradeTimestamp       storage.StorageBackedUint64 // when to do the planned upgrade
	networkFeeAccount      storage.StorageBackedAddress
	l1PricingState         *l1pricing.L1PricingState
	l2PricingState         *l2pricing.L2PricingState
	retryableState         *retryables.RetryableState
	addressTable           *addressTable.AddressTable
	chainOwners            *addressSet.AddressSet
	sendMerkle             *merkleAccumulator.MerkleAccumulator
	blockhashes            *blockhash.Blockhashes
	chainId                storage.StorageBackedBigInt
	chainConfig            storage.StorageBackedBytes
	genesisBlockNum        storage.StorageBackedUint64
	infraFeeAccount        storage.StorageBackedAddress
	brotliCompressionLevel storage.StorageBackedUint64 // brotli compression level used for pricing
	backingStorage         *storage.Storage
	Burner                 burn.Burner
	myNumber               storage.StorageBackedUint64 // this is what we added
}
```

> Next you need to define the offset of your newly added state (note please add it to the end so it won't affect other states):

Reviewer 2: ```suggestion
Next, define the offset of your newly added state (tip: add it to the end so it won't affect other states):
```

```
const (
	versionOffset Offset = iota
	upgradeVersionOffset
	upgradeTimestampOffset
	networkFeeAccountOffset
	chainIdOffset
	genesisBlockNumOffset
	infraFeeAccountOffset
	brotliCompressionLevelOffset
	myNumberOffset // define the offset of your new state here
)
```

> Then, initialize the state under method [OpenArbosState][arbosstate_openstate] and [InitializeArbosState][arbosstate_initializeArbosState]:

Reviewer 2: ```suggestion
Then, initialize the state under the [OpenArbosState][arbosstate_openstate] and [InitializeArbosState][arbosstate_initializeArbosState] methods:
```

OpenArbosState:

```
return &ArbosState{
		arbosVersion,
		backingStorage.OpenStorageBackedUint64(uint64(upgradeVersionOffset)),
		backingStorage.OpenStorageBackedUint64(uint64(upgradeTimestampOffset)),
		backingStorage.OpenStorageBackedAddress(uint64(networkFeeAccountOffset)),
		l1pricing.OpenL1PricingState(backingStorage.OpenSubStorage(l1PricingSubspace)),
		l2pricing.OpenL2PricingState(backingStorage.OpenSubStorage(l2PricingSubspace)),
		retryables.OpenRetryableState(backingStorage.OpenSubStorage(retryablesSubspace), stateDB),
		addressTable.Open(backingStorage.OpenSubStorage(addressTableSubspace)),
		addressSet.OpenAddressSet(backingStorage.OpenSubStorage(chainOwnerSubspace)),
		merkleAccumulator.OpenMerkleAccumulator(backingStorage.OpenSubStorage(sendMerkleSubspace)),
		blockhash.OpenBlockhashes(backingStorage.OpenSubStorage(blockhashesSubspace)),

> 		backingStorage.OpenStorageBackedBigInt(uint64(chainIdOffset)),

Reviewer 2: Qq - thoughts on eliminating all but the most relevant lines of code from snippets, to make it easy for devs to identify, copy, and paste the bits they need? (possibly using `//...` syntax to indicate surrounding code)

Author: Sure, should we also do this at other code blocks?

Reviewer 2: Yeah that would be awesome! This isn't blocking feedback, just a nice pattern to apply wherever it feels appropriate (can be applied incrementally of course) ^^

		backingStorage.OpenStorageBackedBytes(chainConfigSubspace),
		backingStorage.OpenStorageBackedUint64(uint64(genesisBlockNumOffset)),
		backingStorage.OpenStorageBackedAddress(uint64(infraFeeAccountOffset)),
		backingStorage.OpenStorageBackedUint64(uint64(brotliCompressionLevelOffset)),
		backingStorage,
		burner,
		backingStorage.OpenStorageBackedUint64(uint64(myNumberOffset)), // define your new state here
	}, nil
```

InitializeArbosState:

```
	_ = sto.SetUint64ByUint64(uint64(versionOffset), 1) // initialize to version 1; upgrade at end of this func if needed
	_ = sto.SetUint64ByUint64(uint64(upgradeVersionOffset), 0)
	_ = sto.SetUint64ByUint64(uint64(upgradeTimestampOffset), 0)
	_ = sto.SetUint64ByUint64(uint64(myNumberOffset), 0) // initialize your new state around here
```

> You also need to define some methods in this file to read and write this state `myNumber`:

Reviewer 2: ```suggestion
Next, define your getter and setter::
```

```
func (state *ArbosState) SetNewMyNumber(
	newNumber uint64,
) error {
	return state.myNumber.Set(newNumber)
}

func (state *ArbosState) GetMyNumber() (uint64, error) {
	return state.myNumber.Get()
}
```

> After that we need to back to [precompiles directory][precompile_impl_dir_link] and create a new `ArbHi.go` as we introduced in Option 2.

Reviewer 2: ```suggestion
Next, head back to the [precompiles directory][precompile_impl_dir_link] and create a new `ArbHi.go` (introduced in Option 2).
```

> But this time we will add two more new methods to read and write the arbos state:

Reviewer 2: ```suggestion
This time, we'll add two new methods to read and write the ArbOS state:
```

```
package precompiles

// ArbGasInfo provides insight into the cost of using the rollup.
type ArbHi struct {
	Address addr // 0x11a, for example
}

func (con *ArbHi) SayHi(c ctx, evm mech) (string, error) {
	return "hi", nil
}

func (con *ArbHi) GetNumber(c ctx, evm mech) (uint64, error) {
	return c.State.GetMyNumber()
}

func (con *ArbHi) SetNumber(c ctx, evm mech, newNumber uint64) error {
	return c.State.SetNewMyNumber(newNumber)
}
```

> After that, follow the Option 2 guide and add this new precompile contract then run your node.

Reviewer 2: ```suggestion
Follow the procedure detailed in Option 2 in order to add this new precompile contract, and then run your node.
```

[precompile_impl_dir_link]: https://github.com/OffchainLabs/@nitroRepositorySlug@/blob/@nitroVersionTag@/@nitroPathToPrecompiles@/
[precompile_interface_dir_link]: https://github.com/OffchainLabs/@nitroContractsRepositorySlug@/blob/@nitroContractsCommit@/@nitroContractsPathToPrecompilesInterface@/
[precompile_main_file_link]: https://github.com/OffchainLabs/@nitroRepositorySlug@/blob/@nitroVersionTag@/@nitroPathToPrecompiles@/precompile.go
[arbosstate_file_link]: https://github.com/OffchainLabs/@nitroRepositorySlug@/blob/@nitroVersionTag@/@nitroPathToArbosState@/arbosstate.go

> [arbosstate_link]: https://github.com/OffchainLabs/nitro/blob/v2.2.0/arbos/arbosState/arbosstate.go#L38

Reviewer: Can you use @nitroRepositorySlug@ in these 3 lines too?

Author: I didn't apply those tags to there 3 lines is because they point to specific line, so I use a constant url here, if we use it, the file will update to the latest version which might be different line in the future

Reviewer: Alright, in the future maybe we can generalize a method like the one we use for finding the right line of the precompiles functions => https://github.com/OffchainLabs/arbitrum-docs/blob/master/website/src/scripts/precompile-reference-generator.ts#L69
